### PR TITLE
Adds support for new type of source-sink and boundary condition file

### PR DESCRIPTION
**Pull request by bishtgautam**
This PR extends the support of source-sink and boundary conditions in RDycore.

1. Supported types of rainfall datasets include:
   1. CONSTANT: Spatiotemporally constant value
   2. HOMOGENEOUS: Spatially-homogeneous, temporally-varying
   3. RASTER: Spatiotemporally varying in raster format
   4. UNSTRUCTURED: Spatiotemporally varying in an unstructured grid format

2. Supported types of boundary conditions datasets include:
   1. HOMOGENEOUS: Spatially-homogeneous, temporally-varying
   2. UNSTRUCTURED: Spatiotemporally varying in an unstructured grid format

3. A new test is added for the new features added in this PR.

### CODE REVIEW

**Comment by jeff-cohere** [comment]
Are raster-driven simulations going to use HOMOGENEOUS boundary conditions for now?

File: driver/main.c [hunk 1]
```diff
@@ -50,25 +47,72 @@ typedef struct {
...
-  HomogeneousRainData   homogeneous;
-  HeterogeneousRainData heterogeneous;
-} Rain;
-
-// open a Vec that contains data in the following format:
-//
-// time_1 value_1
-// time_2 value_2
-// time_3 value_3
+  DatasetType         type;
```
Comment (jeff-cohere) [inline driver/main.c:85]: We should discuss union data types when you're bored and have a moment. They are a more formal mechanism for having a "variant" type (in other words: a type that holds exactly one value belonging to a set of different possible types). But this is fine for now.

File: driver/main.c [hunk 2]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
+  PetscCall(VecDestroy(&vec));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Sets up the mapping between the unstructured grid dataset cells and the
+///        local cells (for source-sink condition) or boundary edges (for boundary condition).
+/// @param data Pointer to a UnstructuredDataset struct
+/// @return PETSC_SUCESS on success
+static PetscErrorCode CreateUnstructuredDatasetMapping(UnstructuredDataset *data) {
```
Comment (jeff-cohere) [inline driver/main.c:391]: If the unstructured dataset itself is a mapping, would it be accurate to rename this function `CreateUnstructuredDataset`?

File: driver/main.c [hunk 3]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
+    }
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Closes the currently open Vec and loads in the next binary file in the Vec
+/// @param data A pointer to UnstructuredDataset
+/// @return PETSC_SUCESS on success
+static PetscErrorCode OpenANewUnstructuredDataset(UnstructuredDataset *data) {
```
Comment (jeff-cohere) [inline driver/main.c:457]: What do you think about the name `NextUnstructuredDataset`? Too informal?

File: driver/main.c [hunk 4]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
+
+  data->ndata_file++;
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Closes the currently open Vec and free up memory used for mapping
+/// @param data A pointer to UnstructuredDataset struct
+/// @return PETSC_SUCESS on success
+static PetscErrorCode DestroyUnstruturedcDataset(UnstructuredDataset *data) {
```
Comment (jeff-cohere) [inline driver/main.c:489]: There's a typo here. The `c` needs to be moved further forward.

File: driver/main.c [hunk 5]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Set spatially varying BC conditions from an unstructured dataset
+/// @param data        [in]  Pointer to a UnstructuredDataset struct
+/// @param cur_time    [in]  Current time
+/// @param data_values [out] BC values for boundary cells
+/// @return PETSC_SUCESS on success
+PetscErrorCode SetUnstructuredDataset(UnstructuredDataset *data, PetscReal cur_time, PetscReal *data_values) {
```
Comment (jeff-cohere) [inline driver/main.c:510]: Maybe we can call this `SetUnstructuredData` to emphasize that the data is being taken _from_ the dataset and placed into `data_values`.

File: driver/main.c [hunk 6]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
+    }
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief close the currently open raster dataset file and open a new file
+/// @param data [inout] Pointer to a RasterDataset
+/// @return PETSC_SUCESS on success
+static PetscErrorCode OpenANewRasterDataset(RasterDataset *data) {
```
Comment (jeff-cohere) [inline driver/main.c:540]: If you like `NextUnstructureDataSet`, consider `NextRasterDataset` here.

File: driver/main.c [hunk 7]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
-// for each local RDycore grid cell, find the nearest neighbor cell in the rainfall dataset
-static PetscErrorCode SetupHeterogeneousRainDataMapping(RDy rdy, HeterogeneousRainData *hetero_rain) {
+/// @brief For each local RDycore grid cell, find the nearest neighbor cell in the raster dataset
+/// @param rdy   [in]    A RDy struct
+/// @param *data [inout] A pointer to a RasterDataset struct
+/// @return PETSC_SUCESS on success
+static PetscErrorCode CreateRasterDatasetMapping(RDy rdy, RasterDataset *data) {
```
Comment (jeff-cohere) [inline driver/main.c:578]: Consider `CreateRasterDataset`.

File: driver/main.c [hunk 8]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
-// set spatially heterogeneous rainfall rate
-PetscErrorCode SetHeterogeneousRainfall(HeterogeneousRainData *hetero_rain, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
+/// @brief Set rainfall rate for local cells from a spatially raster rainfall dataset. It is assumed that
+///        rainfall rate is mm/hr.
+/// @param data     [in]  A pointer to RasterDataset struct
+/// @param cur_time [in]  Current time
+/// @param ncells   [in]  Number of local cells
+/// @param *rain    [out] Rainfall rate for local cells
+/// @return PETSC_SUCESS on success
+PetscErrorCode SetRasterDataset(RasterDataset *data, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
```
Comment (jeff-cohere) [inline driver/main.c:637]: Consider `SetRasterData`.

File: driver/main.c [hunk 9]
```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
...
 
-// set spatially homogeneous rainfall rate for all grid cells
-PetscErrorCode SetHomogeneousRainfall(HomogeneousRainData *homogeneous_rain, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
+/// @brief Sets rainfall value for local grid cells based on spatially homogenous, temporally-varying rainfall dataset
+/// @param homogeneous_rain [in]  A pointer to HomogeneousDataset
+/// @param cur_time         [in]  Current time
+/// @param ncells           [in]  Number of local cells
+/// @param *rain            [out] Rainfall rate for local cells
+/// @return PETSC_SUCESS on success
+PetscErrorCode SetHomogeneousRainfall(HomogeneousDataset *homogeneous_rain, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
```
Comment (jeff-cohere) [inline driver/main.c:662]: We could use `SetHomogeneousData` here, analogous to `Set{Unstructured,Raster}Data` above.

**Review by jeff-cohere — APPROVED** [review]
This looks fine to me, though there's a lot of logic interspersed in various places that depends on what kind of dataset we're using. Since you've effectively defined a uniform interface for reading data in from files, regardless of the type of data, we could reorganize and encapsulate this logic in a single `RainDataset` variant type that "does the right thing" based on the underlying type of data. I don't think it would take a lot of work for me to do this in a subsequent PR, if you're interested.

**Comment by bishtgautam** [comment]
> Are raster-driven simulations going to use HOMOGENEOUS boundary conditions for now?

The simulation using the raster rainfall dataset will use open outflow boundary.

> I don't think it would take a lot of work for me to do this in a subsequent PR, if you're interested.

A refactor of this code in a future PR would be great. This PR adds the capability required to run simulations for the mid-Atlantic and ROS cases, but one more PR might be needed for the planned simulation in the Amazon basin. Let me check if additional enhancement is needed, and you can then refactor all of the code in one go.

**Comment by bishtgautam** [comment]
I didn't rename `CreateUnstructuredDatasetMapping` to `CreateUnstructuredDataset` because the function creates a mapping from the unstructured grid of the data to the RDycore's cells (for source-sink) or boundary edges (for BCs). Maybe in the future refactor, the logic of `CreateUnstructuredDatasetMapping` can be consumed within `OpenUnstructuredDataset()` (which itself could be renamed as `CreateUnstructuredData`).

**Review by donghuix — APPROVED** [review]
The test simulation with simons domain with the PR looks reasonable.
